dashboard/modules/runtime_env/runtime_env_agent.py

```diff
@@ -102,17 +123,50 @@ def run_setup_with_logger():
                 per_job_logger.debug(f"Worker has resource :"
                                      f"{allocated_resource}")
                 context = RuntimeEnvContext(env_vars=runtime_env.env_vars())
-                self._pip_manager.setup(
-                    runtime_env, context, logger=per_job_logger)
-                self._conda_manager.setup(
-                    runtime_env, context, logger=per_job_logger)
-                self._py_modules_manager.setup(
-                    runtime_env, context, logger=per_job_logger)
-                self._working_dir_manager.setup(
-                    runtime_env, context, logger=per_job_logger)
                 self._container_manager.setup(
                     runtime_env, context, logger=per_job_logger)
 
+                for (manager, uri_cache) in [(self._working_dir_manager,
+                                              self._working_dir_uri_cache),
+                                             (self._conda_manager,
+                                              self._conda_uri_cache),
+                                             (self._pip_manager,
+                                              self._pip_uri_cache)]:
+                    uri = manager.get_uri(runtime_env)
+                    if uri is not None:
+                        if uri not in uri_cache:
+                            size_bytes = manager.create(
+                                uri,
+                                runtime_env,
+                                context,
+                                logger=per_job_logger)
+                            uri_cache.add(
+                                uri, size_bytes, logger=per_job_logger)
+                        else:
+                            uri_cache.mark_used(uri, logger=per_job_logger)
```

Comment:
can we add log statements for cache hits & cache misses here? can be DEBUG level if we think it'll be spammy, but it'll help debugging a lot